packages/api/src/controllers/generate.test.ts

```diff
@@ -190,4 +224,113 @@ describe("controllers/generate", () => {
       });
     }
   });
+
+  describe("request log", () => {
+    async function getLastLog(userId: string) {
+      const [logs] = await db.aiGenerateLog.find(
+        { userId },
+        {
+          limit: 1,
+          order: "coalesce((data->>'startedAt')::bigint, 0) desc",
+        },
+      );
+      expect(logs).toHaveLength(1);
+      return logs[0];
+    }
+
+    it("should log all requests to db", async () => {
+      const res = await client.post("/beta/generate/text-to-image", {
+        prompt: "a man in a suit and tie",
+      });
+      expect(res.status).toBe(200);
+
+      // wait for background operation to write log
+      await sleep(100);
+
+      const log = await getLastLog(nonAdminUser.id);
+      expect(log).toEqual({
+        id: expect.any(String),
+        startedAt: expect.any(Number),
+        durationMs: expect.any(Number),
+        userId: nonAdminUser.id,
+        projectId: nonAdminUser.defaultProjectId,
+        type: "text-to-image",
+        request: {
+          model_id: "SG161222/RealVisXL_V4.0_Lightning", // default gets added
+          prompt: "a man in a suit and tie",
+        },
+        statusCode: 200,
+        response: {
+          message: "success",
+          reqContentType: "application/json",
+        },
+        success: true,
+        error: undefined,
+      } as Required<AiGenerateLog>);
+      expect(log.durationMs).toBeGreaterThanOrEqual(10); // test gateway sleeps for 10ms
+    });
+
+    it("should log failed requests to db", async () => {
+      mockFetchHttpError(
+        500,
+        "application/json",
+        `{"details":{"msg":"sudden error"}}`,
+      );
+
+      const res = await client.post("/beta/generate/text-to-image", {
+        prompt: "a man in a suit and tie",
+      });
+      expect(res.status).toBe(500);
+      expect(await res.json()).toEqual({
+        errors: ["Failed to generate text-to-image"], // 5xx errors get hidden from end users
+      });
+
+      await sleep(100);
```

Comment:
I think it is theoretically possible for this test to fail yeah. But we really only need this because we save the log with a `setImmediate` instead of synchronously. `setImediate` executes as soon as possible on the node event loop. Then it's going to save on the DB on localhost, which also shouldn't take more than a couple `ms`.

I could make this into a loop tho if you think there's still a risk.

Reply:
Not sure to be honest. I'm somewhere on a fence, because:
- on one hand => I think we should use the pattern of retries everywhere, because if we have a lot of these, then at some point we'll start having flaky test (and it's a pain, we already have them in go-livepeer, catalyst-api, and catalyst)
- on the other hand => we have these in many places, so we can leave it for now and take the reactive approach, so leave it as it as and maybe it won't happen in practice